lib/fir/statements.h

```diff
@@ -0,0 +1,231 @@
+// Copyright (c) 2018, NVIDIA CORPORATION.  All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef FORTRAN_IR_STATEMENTS_H_
+#define FORTRAN_IR_STATEMENTS_H_
+
+#include "mixin.h"
+#include "../common/idioms.h"
+#include "../common/indirection.h"
+#include "../evaluate/expression.h"
+#include "../parser/parse-tree.h"
+#include "../semantics/symbol.h"
+#include <initializer_list>
+#include <iostream>
+
+namespace Fortran::IR {
+
+class Statement;
+class BasicBlock;
+CLASS_TRAIT(StatementTrait)
+CLASS_TRAIT(TerminatorTrait)
+CLASS_TRAIT(ActionTrait)
+
+using Expression = const common::OwningPointer<evaluate::GenericExprWrapper>;
+using Variable = const semantics::Symbol;
+using PathVariable = const parser::Variable;
+
+struct Evaluation
+  : public SumTypeCopyMixin<
+        std::variant<Expression *, Variable *, PathVariable *>> {
+  SUM_TYPE_COPY_MIXIN(Evaluation)
+  Evaluation(PathVariable *pv) : SumTypeCopyMixin{pv};
+  template<typename A> Evaluation(A *a) : SumTypeCopyMixin{a} {}
+  std::string dump() const;
+};
+
+struct Stmt_impl {
```

Comment:
Does `_impl` mean the struct is not meant to be part of the API? Maybe add an explanation to `C++style.md`.

Reply:
These are for abstract classes for ad hoc polymorphism and a place for shared code.